client/apps.go

```diff
@@ -261,3 +269,105 @@ func (client *Client) Restart(names []string, opts RestartOptions) (changeID str
 	}
 	return client.doAsync("POST", "/v2/apps", nil, nil, bytes.NewReader(buf))
 }
+
+func (app *AppInfo) Notes() string {
+	if !app.IsService() {
+		return "-"
+	}
+
+	var notes = make([]string, 0, 2)
+	var seenTimer, seenSocket bool
+	for _, act := range app.Activators {
+		switch act.Type {
+		case "timer":
+			seenTimer = true
+		case "socket":
+			seenSocket = true
+		}
+	}
+	if seenTimer {
+		notes = append(notes, "timer-activated")
+	}
+	if seenSocket {
+		notes = append(notes, "socket-activated")
+	}
+	if len(notes) == 0 {
+		return "-"
+	}
+	return strings.Join(notes, ",")
+}
+
+func AppInfosFromSnapAppInfos(apps []*snap.AppInfo) []AppInfo {
+	// TODO: pass in an actual notifier here instead of null
+	//       (Status doesn't _need_ it, but benefits from it)
+	sysd := systemd.New(dirs.GlobalRootDir, progress.Null)
+
+	out := make([]AppInfo, 0, len(apps))
+	for _, app := range apps {
+		appInfo := AppInfo{
+			Snap:     app.Snap.InstanceName(),
+			Name:     app.Name,
+			CommonID: app.CommonID,
+		}
+		if fn := app.DesktopFile(); osutil.FileExists(fn) {
+			appInfo.DesktopFile = fn
+		}
+
+		appInfo.Daemon = app.Daemon
+		if !app.IsService() || !app.Snap.IsActive() {
+			out = append(out, appInfo)
+			continue
+		}
+
+		// collect all services for a single call to systemctl
+		serviceNames := make([]string, 0, 1+len(app.Sockets)+1)
+		serviceNames = append(serviceNames, app.ServiceName())
+
+		sockSvcFileToName := make(map[string]string, len(app.Sockets))
+		for _, sock := range app.Sockets {
+			sockUnit := filepath.Base(sock.File())
+			sockSvcFileToName[sockUnit] = sock.Name
+			serviceNames = append(serviceNames, sockUnit)
+		}
+		if app.Timer != nil {
+			timerUnit := filepath.Base(app.Timer.File())
+			serviceNames = append(serviceNames, timerUnit)
+		}
+
+		// sysd.Status() makes sure that we get only the units we asked
+		// for and raises an error otherwise
+		sts, err := sysd.Status(serviceNames...)
+		if err != nil {
+			logger.Noticef("cannot get status of services of app %q: %v", app.Name, err)
```

Comment:
Since it's in the `client` now, and we don't seem to be using logger in this way anywhere in the package, perhaps we could just error out here?

Alternatively, could use something like we do in the `store` and return whatever was successfuly collected and an error (or nil). The message can be  something like `failed to get status of the follwoing services: snap.foo.bar.service, snap.bar.baz.service`

Reply:
the error is rare enough that just failing is ok i reckon (and if systemd is having a bad day, your day is not going to be much better).